### PR TITLE
Separate middleware options

### PR DESCRIPTION
**Pull request by hananiel**
## Description

The goal of this PR is to realize the pattern we discussed in this comment: https://github.com/SteeltoeOSS/Steeltoe/pull/1120#issuecomment-1540458522

Notable changes: 
1. Remove the various overloads of `EndpointMiddleWare`, `EndpointMiddleware<TResult>`, `EndpointMiddleware <TArgument, TResult>` with just `EndpointMiddleware<TArgument, TResult>`. 
2. Implement Async 

An attempt was made to separate transport related options entirely but I was only able to achieve partial separation. The transport related options are the common `HttpMiddlewareOptions`(Formerly `EndpointOptionsBase`). Specific endpoint handlers use an options class derived from this for example, `MetricEndpointOptions: HttpMiddlewareOptions `and contain other values. The key thing to note is where these options are loaded from `management:endpoint:options:metics` etc. I think it makes sense to not have different unrelated classes for transport and non-transport options since we cannot change the shape of the configuration settings.

Note that the `ManagementEndpointOptions` is actually related to the different contexts we serve endpoints in for example: /actuator/* (default path) context is served by named option of name "actuator" and type `ManagementEndpointOptions`. And the context /cloudfoundryapplication/* is served by named option of name "CloudFoundry" and same type `ManagementEndpointOptions`. I am open to a different name or different design here. I tried to make it extensible so future contexts would involve just adding another named option but it is not strictly needed.  

Fixes #{bug number}

## Quality checklist

- [ ] Your code complies with our [Coding Style](https://github.com/SteeltoeOSS/.github/blob/main/contributing-docs/contributing-code-style.md).
- [ ] You've updated unit and/or integration tests for your change, where applicable.
- [ ] You've updated documentation for your change, where applicable.
      If your change affects other repositories, such as [Documentation](https://github.com/SteeltoeOSS/Documentation), [Samples](https://github.com/SteeltoeOSS/Samples) and/or [MainSite](https://github.com/SteeltoeOSS/MainSite), add linked PRs here.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.
- [ ] You've added required license files and/or file headers (explaining where the code came from with proper attribution), where code is copied from StackOverflow, a blog, or OSS.

### CODE REVIEW

**Review by bart-vmware — CHANGES_REQUESTED** [review]
I'm pretty happy with where this is going. IoC registrations are more straightforward now, and duplicate serialization logic is gone. This gets pretty close to what I proposed at https://github.com/SteeltoeOSS/Steeltoe/pull/1120#issuecomment-1540458522, thanks for trying to make it fit as well as possible.

What I don't understand though: This PR is named "Separate middleware options", but I'm not seeing any actual separation between transport-specific options and transport-agnostic options. It appears this PR changes the following:

```
interface IEndpointHandler (was: IEndpoint)
	property: HttpMiddlewareOptions (was: IEndpointOptions/EndpointOptionsBase)

class EndpointMiddleware<>
	property: ManagementEndpointOptions (unchanged)
```

I'm confused by these names. Do you intend to implement the separation after this PR is merged into #1120? A bit more info in the PR description would be welcome.

File: src/Management/src/Endpoint/Mappings/IMappingsEndpointHandler.cs
Comment (TimHess) [reply src/Management/src/Endpoint/Mappings/IMappingsEndpointHandler.cs:7]: The `Mappings` endpoint has [Spring](https://docs.spring.io/spring-boot/docs/3.1.x/reference/html/actuator.html#actuator) roots (described there as `Displays a collated list of all @RequestMapping paths.`), I don't know the full history, but I wouldn't be surprised if it predated OpenAPI. The naming of this endpoint should probably stay aligned with the front ends that support it (eg: Apps Manager, Spring Boot Admin, App Live View). Would adding some (XML) comments help?

File: src/Management/src/Endpoint/Env/EnvironmentEndpointHandler.cs
Comment (TimHess) [reply src/Management/src/Endpoint/Env/EnvironmentEndpointHandler.cs:132]: Should we run the code cleanup to resolve here and now, or do you think we should wait until this PR is merged into the other?

File: src/Management/src/Endpoint/HeapDump/HeapDumpEndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/HeapDump/HeapDumpEndpointMiddleware.cs:21]: The middleware matches the EndpointHandler TArg, TResult. We handle the dump collection and the resulting file in the endpointhandler and then we do the compression, and transport related things in the middleware. We could call it object and then cast to string for the endpointHandler's result - but that seems odder than passing an empty string. The other alternative is convert the stream to string but that seems inefficient than writing to the response body directly.

File: src/Management/src/Endpoint/Loggers/LoggersEndpointHandler.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Loggers/LoggersEndpointHandler.cs:42]: Yes sometimes the `IDynamicLoggerProvider` is not available in the service container.

Comment (TimHess) [reply src/Management/src/Endpoint/Loggers/LoggersEndpointHandler.cs:42]: It looks like we have code to handle it being null at request time, but would it be better if we could handle that concern in the constructor since this endpoint is (I think?) useless without it?

Comment (hananiel) [reply src/Management/src/Endpoint/Loggers/LoggersEndpointHandler.cs:42]: good point, I was thinking of the IDynamicMessageProcessor.

File: src/Messaging/src/Messaging/Handler/HandlerMethod.cs
Comment (TimHess) [reply src/Messaging/src/Messaging/Handler/HandlerMethod.cs]: Beans observation is a good one, but out of scope for this PR

File: src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs]: changed to `protected virtual JsonSerializerOptions GetSerializerOptions()`

File: src/Management/src/Endpoint/CloudFoundry/SecurityUtils.cs
Comment (hananiel) [reply src/Management/src/Endpoint/CloudFoundry/SecurityUtils.cs:19]: In this case, I feel the class name helps someone trouble shoot faster than mentioning the now abstract "Endpoint"

File: src/Management/src/Endpoint/Options/ConfigureEndpointOptions.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Options/ConfigureEndpointOptions.cs:12]: done.

File: src/Management/src/Endpoint/HeapDump/HeapDumpEndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/HeapDump/HeapDumpEndpointMiddleware.cs:21]: @TimHess  thoughts here?

File: src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs]: removed.

File: src/Management/src/Endpoint/CloudFoundry/SecurityUtils.cs
Comment (TimHess) [reply src/Management/src/Endpoint/CloudFoundry/SecurityUtils.cs:19]: I'm not deep in the context on this one.. Is it likely to be seen/used by a Steeltoe developer or consumer? Are we going to expose Steeltoe users to the handlers, or are they really only an internal detail? It might be simpler if our users aren't aware there are things called handlers at all

File: src/Management/src/Endpoint/Metrics/MetricsEndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Metrics/MetricsEndpointMiddleware.cs:123]: done

File: src/Management/src/Endpoint/Loggers/LoggersEndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Loggers/LoggersEndpointMiddleware.cs:19]: done

File: src/Management/src/Endpoint/Loggers/LoggersEndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Loggers/LoggersEndpointMiddleware.cs:65]: done.

File: src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs:31]: I tried but this is not possible because the IEndpointMiddleware does not have (and cannot have the EndpointHandler<TRequest, TResult> and we need the options at in the ActuatorEndpointMapper for each IEndpointMiddleware.

File: src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Middleware/EndpointMiddleware.cs:63]: Done.

File: src/Management/src/Endpoint/Properties/AssemblyInfo.cs
Comment (TimHess) [inline src/Management/src/Endpoint/Properties/AssemblyInfo.cs:11]:
```suggestion
```

File: src/Management/test/Kubernetes.Test/HostBuilderExtensionsTest.cs
Comment (hananiel) [reply src/Management/test/Kubernetes.Test/HostBuilderExtensionsTest.cs:23]: No it's not a breaking change, but we did change how the tests worked.

File: src/Management/src/Endpoint/Options/ConfigureEndpointOptions.cs
Comment (TimHess) [reply src/Management/src/Endpoint/Options/ConfigureEndpointOptions.cs:12]: `HttpMiddlewareOptions` doesn't appear to be abstract, am I missing something?

File: src/Management/src/Endpoint/Metrics/MetricsEndpointMiddleware.cs
Comment (TimHess) [reply src/Management/src/Endpoint/Metrics/MetricsEndpointMiddleware.cs]: > The same applies to `HealthEndpointMiddleware` and `LoggersEndpointMiddleware`
I don't see the same change applied in those yet

File: src/Management/src/Endpoint/Mappings/IMappingsEndpointHandler.cs
Comment (hananiel) [reply src/Management/src/Endpoint/Mappings/IMappingsEndpointHandler.cs:8]: Lets add xml comment about what this is in Dotnet terms for example RouteMappings (instead of changing the code)

File: src/Management/test/Kubernetes.Test/HostBuilderExtensionsTest.cs
Comment (hananiel) [reply src/Management/test/Kubernetes.Test/HostBuilderExtensionsTest.cs:23]: To give more context, I believe the  change happened when we added the `ActuatorEndpointMapper` which happened in earlier PR, but I think we fixed the test here. Not a 100% sure.

File: src/Management/src/Endpoint/Mappings/IMappingsEndpointHandler.cs
Comment (TimHess) [reply src/Management/src/Endpoint/Mappings/IMappingsEndpointHandler.cs:8]: I see a note added on the middleware. If we need more clarity elsewhere, we can add that in the future

**Review by TimHess — APPROVED** [review]
LGTM

**Pull request merged**
